### PR TITLE
Fix updating editor when editing selected text

### LINKED ISSUE #1509: Error thrown when typing on selected text
**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

draft-js throws `cannot read property 'update' of undefined` when I type on selected text. To reproduce it, go to https://draftjs.org and type:

```
012
012
012
```

Then select the '1' of the second line. Now try to type '1'. Now type '2', it'll work. Now if you select the '2' that was just inserted and try to type '1' again, it'll not update.

Here's a gif demonstrating the bug:

screen

* I can reproduce it on https://draftjs.org/ and in my projects too.
* I can reproduce it with other strings too, like `abc` instead of `012`

**What is the expected behavior?**

No errors thrown and the text should be updated correctly.

**Which versions of Draft.js, and which browser / OS are affected by this issue? Did this work in previous versions of Draft.js?**

draft-js: 0.10.4
OS: Windows 10
Browser: Chrome 62, Firefox 56

I don't know if it worked on previous versions of draft-js. I'm a new user 😄

### PR DESCRIPTION
**Pull request by existentialism**
Fixes #1509.

Ref: https://github.com/facebook/draft-js/pull/719

### CODE REVIEW

**Comment by mitermayer** [comment]
Thanks for this! @existentialism  do you think you could also add a test for it ? :)

**Comment by mitermayer** [comment]
adding this PR to my list of PR's to test/import

**Comment by bradbirnbaum** [comment]
I'm bumping into this issue - any ETA on when this will be published?   Thanks for your hard work!

**Comment by existentialism** [comment]
@mitermayer sorry for delay, in the midst of preppin' for turkey day :) will try to get a test(s) written asap!

**Comment by juliankrispel** [comment]
Thanks for fixing this! Looks like like this is good to merge @mitermayer?

**Comment by sontek** [comment]
Any chance this can get merged + release?   This is causing us issues in prod :)

**Comment by juliankrispel** [comment]
@sontek @existentialism this will get merged and releaed evtl. but just fyi the release process is a lot slower in this repository due to necessity.

If you want to use it now, fork from this commit and use that until it's merged. Sry for the inconvenience, it's just that with a repository this much used it's hard to move fast :(

File: src/component/handlers/edit/__tests__/editOnBeforeInput.test.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+const DEFAULT_SELECTION = {
+  anchorKey: 'a',
+  anchorOffset: 0,
+  focusKey: 'a',
+  focusOffset: 0,
+  isBackward: false,
+};
+
+//const collapsedSelection = new SelectionState(DEFAULT_SELECTION);
```
Comment (flarnie) [inline src/component/handlers/edit/__tests__/editOnBeforeInput.test.js]: Did we mean to remove this?

File: src/component/handlers/edit/__tests__/editOnBeforeInput.test.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    update: jest.fn(),
+  };
+
+  onBeforeInput(editor, getInputEvent('A'));
+
+  expect(editor.update).toHaveBeenCalledTimes(1);
+
+  const newEditorState = editor.update.mock.calls[0][0];
+  expect(newEditorState.getCurrentContent()).toMatchSnapshot();
+});
```
Comment (flarnie) [inline src/component/handlers/edit/__tests__/editOnBeforeInput.test.js]: Hooray~ Great tests!

tests_pass

**Comment by existentialism** [comment]
@flarnie @mitermayer thanks!

**Comment by JLarky** [comment]
I think this issue is still not solved.

Text I use `abc` now If I select `a` and I type `a` it may or may not work depending on how I selected the character: from left to right or from right to left (but mouse or keyboard doesn't matter). Good thing that it's no longer crashes :+1:, but behavior is still wrong.

Can see this bug on both my website and https://draftjs.org/

**Comment by existentialism** [comment]
@JLarky ill take a look

**Pull request closed without merging**